### PR TITLE
Add full training loop

### PR DESCRIPTION
**Pull request by Lucaskabela**
Another one with lots of changes!   A brief list:

- Create consolidated model wrapping a speech and text autoencoder, as well as support for a future discriminator
- Add explicit speech and text loss functions (speech requires also computing loss w.r.t stop tokens)
- Add evaluation code to compute per and losses on holdout set
- Add crossmodel and supervised objectives & training
- Add printing of losses (all 6) on each epoch, as well as validation

### CODE REVIEW

**Comment by Lucaskabela** [comment]
Not tested yet!

File: src/train.py [hunk 1]
```diff
@@ -16,67 +16,248 @@
...
+
+    # Should be [batch_size x seq_length] for stop 
+    pred_loss = F.mse_loss(pred_mel, gold_mel)
 
+    # TODO: return actual lengths, not just computed off padding
+    # currently, find first nonzero (so pad_idx) in pos_mel, or set to length
+    end_mask_max, end_mask_idx = torch.max((pos_mel == PAD_IDX), dim=1)
+    end_mask_idx[end_mask_max == 0] = pos_mel.shape[1] - 1
+    stop_label = F.one_hot(end_mask_idx, pos_mel.shape[1]).float()
+    stop_loss = F.binary_cross_entropy_with_logits(stop_pred.squeeze(), stop_label)
```
Comment (billyang98) [inline src/train.py:97]: nit: need to return the loss?

**Review by billyang98 — APPROVED** [review]

File: src/train.py (hunk 1 above)
Comment (liaojh1998) [reply src/train.py:97]: I think so for both losses, and maybe these need to be weighted such that `pred_loss` gains more weight than `stop_loss` like in the paper.

File: src/train.py [hunk 2]
```diff
@@ -16,67 +16,248 @@
...
+            losses['t_cm'].append(t_cm_loss.detach().cpu().item())
+
+            text_pred = model.asr(None, mel.to(DEVICE), infer=True).squeeze()
+            len_mask_max, len_mask_idx = torch.max((text_pred == PAD_IDX), dim=1)
+            len_mask_idx[end_mask_max == 0] = text_pred.shape[1] - 1
+            print("Lengths", len_mask_idx.shape)
+            per += compute_per(character.to(DEVICE), text_pred, text_len.to(DEVICE), len_mask_idx)
+            n_iters += 1
+
+    model.train()
```
Comment (liaojh1998) [inline src/train.py]: I don't think the `model.train()` should be added here. This should be put in the `train()` function on line 155-158:
```python3
    ...
    for epoch in range(args.epochs):
        model.train()

        # Training model
        supervised_dataloader = DataLoader(supervised_train_dataset, batch_size=args.batch_size, shuffle=True, collate_fn=collate_fn_transformer, drop_last=True, num_workers=16)
        ...
```

Comment (liaojh1998) [reply src/train.py]: On another note, probably leave a comment here for later: `# TODO: probably evaluate speech somehow?`

File: src/train.py [hunk 3]
```diff
@@ -16,67 +16,248 @@
...
+    val_dataset = get_dataset('val.csv')
+    full_train_dataset = get_dataset('full_train.csv')
+
+    # init models and optimizers
+    text_m = TextRNN(args).to(DEVICE)  
+    speech_m = SpeechRNN(args).to(DEVICE)
+    discriminator = None
+    model = UNAST(text_m, speech_m, discriminator)
+    optimizer = torch.optim.Adam(model.parameters(), lr=args.lr,  weight_decay=1e-5)
+    best = 100
```
Comment (liaojh1998) [inline src/train.py]: Would it be possible to create a separate function for initializing the model (and no need to add lots of `args` test to it yet):

```python3
def initialize_model(args, device):
    # We can add schedulers to this later if we need to

    # initialize model
    text_m, speech_m, discriminator = None, None, None
    if args.model_type == 'rnn':
        text_m = TextRNN(args).to(device)
        speech_m = SpeechRNN(args).to(device)
    if args.model_type == 'transformer':
        ...
    if args.use_discriminator:
        discriminator = ...
    model = UNAST(text_m, speech_m, discriminator)

    # initialize optimizer
    optimizer = None
    if args.optim_type == 'adam':
        optimizer = torch.optim.Adam(model.parameters(), lr=args.lr, weight_decay=1e-5)

    return model, optimizer
```

So that we can reuse the `train()` function with other models:

```python3
def train(args, model, optimizer):
    ...
```

And call `initialize_model(args, DEVICE)` in `"__main__"`.

File: src/train.py [hunk 4]
```diff
@@ -16,67 +16,248 @@
...
+    """
+    character, mel, mel_input, pos_text, pos_mel, text_len = batch
+    character, mel, mel_input = character.to(DEVICE), mel.to(DEVICE), mel_input.to(DEVICE)
+    mel_gold, char_gold = mel.detach(), character.detach().permute(1, 0)
+
+    text_pred = model.text_ae(character).permute(1, 2, 0)
+    pred, stop_pred = model.speech_ae(mel, mel_input)
+    
+    # TODO: add end length loss
+    s_ae_loss = speech_loss(mel_gold, pos_mel, pred, stop_pred)
```
Comment (liaojh1998) [inline src/train.py]: Referencing the previous comments, need to return the losses here and maybe weigh them differently.

File: src/train.py [hunk 5]
```diff
@@ -16,67 +16,248 @@
...
+
+def supervised_step(model, batch):
+    character, mel, mel_input, pos_text, pos_mel, text_len = batch
+    character, mel, mel_input = character.to(DEVICE), mel.to(DEVICE), mel_input.to(DEVICE)
+    mel_gold, char_gold = mel.detach(), character.detach().permute(1, 0)
+
+    pred, stop_pred = model.tts(character, mel_input)
+    text_pred = model.asr(character, mel).permute(1, 2, 0)
+    
+    tts_loss = speech_loss(mel_gold, pos_mel, pred, stop_pred)
```
Comment (liaojh1998) [inline src/train.py]: Referencing the previous comments, need to return the losses here and maybe weigh them differently.

File: src/train.py [hunk 6]
```diff
@@ -16,67 +16,248 @@
 import torch.nn as nn
 import torch
 import numpy as np
+from collections import defaultdict
 
 DEVICE = init_device()
```
Comment (liaojh1998) [inline src/train.py]: Not an issue for now, but maybe we can make `device` one of the `args` later one.

File: src/train.py [hunk 7]
```diff
@@ -16,67 +16,255 @@
...
+        discriminator = None
+        model = UNAST(text_m, speech_m, discriminator)
+        optimizer = torch.optim.Adam(model.parameters(), lr=args.lr,  weight_decay=1e-5)
+        s_epoch, best = 0, 100
+    else:
+        model = UNAST(None, None, None)
+        optimizer = toch.optim.Adam(model.parameters(), lr=args.lr, weight_decay=1e-5)
+        data = load_ckp(args.load_path, model, optimizer)
+        s_epoch, best, model, optimizer = data
+        s_epoch = s_epoch + 1
```
Comment (liaojh1998) [inline src/train.py]: Just saw this, probably add a `continue_training` args part to the `initialize_model()` as well.

**Review by liaojh1998 — CHANGES_REQUESTED** [review]
Looks good on adding the big training loop. I have an additional function request, but it can also be added later if we want to run this right now.

File: src/train.py (hunk 6 above)
Comment (Lucaskabela) [reply src/train.py]: Consider it done! I think instead of passing around the device variable everywhere, it is easier to just make it global in the main method.  We could also move it to the top of train if you think it is better suited there!

File: src/train.py (hunk 1 above)
Comment (Lucaskabela) [reply src/train.py:97]: I just returned the sum, following how the loss is computed in Tacotron 2: https://github.com/NVIDIA/tacotron2/blob/master/loss_function.py

However, I think we can definitely add parameters (like lambda_xyz) to weight each of the loss functions in the args.  Any ideas how we might tune these?

File: src/train.py (hunk 2 above)
Comment (Lucaskabela) [reply src/train.py]: Good insight about the `model.train()` call, it definitely makes more sense in the "train" function.

File: src/train.py (hunk 1 above)
Comment (liaojh1998) [reply src/train.py:97]: If it turns out most of the speech output is just an early stop, then we will have to tune this. For now, let's run it and see what happens.

File: src/train.py [hunk 8]
```diff
@@ -199,91 +343,57 @@ def train_speech_auto(args):
...
+            pass
+        model = UNAST(text_m, speech_m, discriminator)
+
+        # initialize optimizer
+        optimizer = None
+        if args.optim_type == 'adam':
+            optimizer = torch.optim.Adam(model.parameters(), lr=args.lr, weight_decay=1e-5)
+            s_epoch, best = 0, 100
+
+    return s_epoch, best, model, optimizer
```
Comment (liaojh1998) [inline src/train.py]: I like the new function, and this is fine for now. We'll define these hyperparameters later.

**Review by liaojh1998 — APPROVED** [review]
Looks good! I like the refactoring. Looking forward to seeing how the model trains!

**Review by billyang98 — APPROVED** [review]

File: src/network.py [hunk 9]
```diff
@@ -149,33 +192,34 @@ def infer_sequence(self, hidden_state, enc_output, enc_ctxt_mask, max_len=1000):
...
 
         while keep_gen:
-            dec_out, stop_pred, hidden_state = self.decode(input_, hidden_state, enc_output, enc_ctxt_mask)
+            (dec_out, stop_pred), hidden_state = self.decode(input_, hidden_state, enc_output, enc_ctxt_mask)
+            stop_pred = stop_pred.squeeze()
             stops.append(stop_pred)
             # set stop_lens here!
             outputs.append(dec_out)
-            input_ = outputs[-1]
+            input_ = outputs[-1].detach().permute(0, 2, 1)
```
Comment (liaojh1998) [inline src/network.py]: If I'm understanding this right, are we removing the gradients on generating the sequence for the cross-modal loss?

That is, if we had the following data flow:
`speech --> seq2seq --> text --> seq2seq --> speech`

We're only propagating the gradients on `text --> seq2seq --> speech` but not for `speech --> seq2seq --> text`?

File: src/network.py [hunk 10]
```diff
@@ -149,33 +192,34 @@ def infer_sequence(self, hidden_state, enc_output, enc_ctxt_mask, max_len=1000):
...
             i += 1
 
             # double check this!
             stop_mask = (stop_pred > .5).logical_and(stop_lens == 0)
             stop_lens[stop_mask] = i
-            keep_gen = torch.all(stop_lens.neq(0)) and i < max_len
+            keep_gen = torch.any(stop_lens.eq(0)) and i < max_len
```
Comment (liaojh1998) [inline src/network.py:228]: Nice catch!

File: src/network.py [hunk 11]
```diff
@@ -194,7 +238,7 @@ def decode_sequence(self, target, hidden_state, enc_output, enc_ctxt_mask, teach
             if random.random() < teacher_ratio and i + 1 < max_out_len:
                 input_ = target[:, i+1, :].unsqueeze(1)
             else:
-                input_ = outputs[-1]
+                input_ = outputs[-1].detach().permute(0, 2, 1)
```
Comment (liaojh1998) [inline src/network.py]: Same here. Why are the gradients detached here?

Also, if I'm understanding this right, are we using only the last output, at time t, of the decoder as input to the decoder for the next step, t + 1, instead of a sequence of outputs from 1 to t?

**Review by liaojh1998 — COMMENTED** [review]
Nice work! I had some questions on architecture but everything else looks good.

File: src/network.py (hunk 9 above)
Comment (liaojh1998) [reply src/network.py]: My bad, I just found out that `.detach()` creates a new tensor sharing values instead of modifying the original one and removing it from the graph. So the loss should propagate here since you add the non-detached one to the list.

File: src/network.py (hunk 11 above)
Comment (liaojh1998) [reply src/network.py]: The comment on detached can be ignored, but the decoding question remains here.

Comment (Lucaskabela) [reply src/network.py]: In the case we are using autoregressive input, we have to detach() the gradient here - if it does not get detached, we get NaN's (I think this is because input is an output of the LSTM, so feeding it back in leads to some error/overlapping gradients?)
